### PR TITLE
  better error handling

### PR DESCRIPTION
**Pull request by berda97**
I put tray and catch in each method,
in the catch I inserted the return status 500

### CODE REVIEW

**Comment by berda97** [comment]
better error handling

File: AccountingService/Controllers/UsersController.cs [hunk 1]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
 
         private NetSalaryService netSalaryService;
 
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove extra spacing

File: AccountingService/Controllers/UsersController.cs [hunk 2]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+
         public UsersController(SalaryConversionContext context) : base()
         {
             salaryConversionContext = context;
             exchangeRateService = new ExchangeRateService();
             netSalaryService = new NetSalaryService();
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove extra spacing

File: AccountingService/Controllers/UsersController.cs [hunk 3]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+                var users = salaryConversionContext.User.ToArray(); // izvuci sve user iz tabele(user) 
+                return Ok(new
+                {
+                    Count = users.Length,
+                    Users = users,
+                });
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Error: " + ex.Message);
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: This `Console.WriteLine` is currently unnecessary. Once we introduce logging, we might need something similar, but for now, please remove it.

File: AccountingService/Controllers/UsersController.cs [hunk 4]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+                return new StatusCodeResult(500);
+            }
         }
 
-
         [HttpGet("{id}")]
         public IActionResult GetUserById(int id)
         {
-            var user = salaryConversionContext.User.Where(u => u.ID == id).SingleOrDefault();
-            if (user == null)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:42]: Why did you remove this check?

File: AccountingService/Controllers/UsersController.cs [hunk 5]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
-            }
+                var user = salaryConversionContext.User.Where(u => u.ID == id).SingleOrDefault();
 
-            return Ok(user);
 
+                return Ok(user);
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Error: " + ex.Message);
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Same as previous comment. Remove this. It is unnecessary.

File: AccountingService/Controllers/UsersController.cs [hunk 6]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
 
-
-
-
         [HttpPost]
         public IActionResult CreateUser([FromBody] User request)
         {
-            salaryConversionContext.User.Add(request);
-            int changes = salaryConversionContext.SaveChanges();
-            if (changes == 0)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:59]: Why did you remove this check?

File: AccountingService/Controllers/UsersController.cs [hunk 7]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+            try
             {
-                return BadGateway();
+                salaryConversionContext.User.Add(request);
+                int changes = salaryConversionContext.SaveChanges();
+                return Ok();
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine("Error: " + ex.Message);
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove this.

File: AccountingService/Controllers/UsersController.cs [hunk 8]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
             }
-            return Ok();
         }
 
         [HttpPut]
         public IActionResult UpdateUser([FromBody] User request)
         {
-            salaryConversionContext.User.Update(request);
-            int changes = salaryConversionContext.SaveChanges();
-            if (changes == 0)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:71]: Why did you remove this check?

File: AccountingService/Controllers/UsersController.cs [hunk 9]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+                salaryConversionContext.User.Update(request);
+                int changes = salaryConversionContext.SaveChanges();
+
+                return Ok();
             }
 
-            return Ok();
+            catch (Exception ex)
+            {
+                Console.WriteLine("Error: " + ex.Message);
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove this.

File: AccountingService/Controllers/UsersController.cs [hunk 10]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
         {
-            var user = salaryConversionContext.User.Where(u => u.ID == id).SingleOrDefault();
-            if (user == null)
             {
-                return NotFound();
-            }
-            salaryConversionContext.User.Remove(user);
-            int changes = salaryConversionContext.SaveChanges();
-
-            if (changes == 0)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:90]: Why did you remove this?

File: AccountingService/Controllers/UsersController.cs [hunk 11]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
         [HttpGet("{id}/salary")]
+
         public IActionResult GetUserSalary(int id, [FromQuery(Name = "currency")] string currency, [FromQuery(Name = "isNetSalary")] bool isNetSalary)
         {
-            double salary = salaryConversionContext.User
-                .Where(u => u.ID == id)
-                .Select(u => u.Salary)
-                .SingleOrDefault();
-
-            if (salary == 0)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:107]: The check is missing.

File: AccountingService/Controllers/UsersController.cs [hunk 12]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
-            });
-
-        }
-
-
-
-        private IActionResult BadGateway()
-        {
-            return StatusCode(StatusCodes.Status502BadGateway);
+                Console.WriteLine("Error: " + ex.Message);
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove this.

File: AccountingService/Controllers/UsersController.cs [hunk 13]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
 
-            double exchangeRate = exchangeRateService.GetCurrencyExchangeRate(currency);
-            salary = salary * exchangeRate;
 
-            if (isNetSalary)
-            {
-                double netSalary = netSalaryService.Calculate(salary);
+
+                double exchangeRate = exchangeRateService.GetCurrencyExchangeRate(currency);
+                salary = salary * exchangeRate;
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Everything after this line can not throw an exception. Move it out of the `try/catch` block.

**Review by ghost — CHANGES_REQUESTED** [review]
There are some changes that must be applied before proceeding.

File: AccountingService/Controllers/UsersController.cs [hunk 14]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
                 {
                     Value = salary,
-                    Currency = currency,
-                    NetSalary = netSalary
+                    Currency = currency
                 });
             }
-
-            return Ok(new
+            catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs [hunk 15]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
+                    {
+                        return NotFound();
+                    }
+                    salaryConversionContext.User.Remove(user);
+                    int changes = salaryConversionContext.SaveChanges();
+
+
+                    return Ok();
+                }
+                catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs [hunk 16]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
             {
-                return BadGateway();
+                salaryConversionContext.User.Update(request);
+                int changes = salaryConversionContext.SaveChanges();
+
+                return Ok();
             }
 
-            return Ok();
+            catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs [hunk 17]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
-            int changes = salaryConversionContext.SaveChanges();
-            if (changes == 0)
+            try
             {
-                return BadGateway();
+                salaryConversionContext.User.Add(request);
+                int changes = salaryConversionContext.SaveChanges();
+                return Ok();
+            }
+            catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs [hunk 18]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
             {
-                return NotFound();
-            }
+                var user = salaryConversionContext.User.Where(u => u.ID == id).SingleOrDefault();
 
-            return Ok(user);
 
+                return Ok(user);
+            }
+            catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs [hunk 19]
```diff
@@ -15,130 +16,155 @@ public class UsersController : ControllerBase
...
-            });
-
+                var users = salaryConversionContext.User.ToArray(); // izvuci sve user iz tabele(user) 
+                return Ok(new
+                {
+                    Count = users.Length,
+                    Users = users,
+                });
+            }
+            catch (Exception ex)
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `ex` is not necessary. It is not used.

File: AccountingService/Controllers/UsersController.cs (hunk 1 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs]: done

File: AccountingService/Controllers/UsersController.cs (hunk 2 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs]: done

File: AccountingService/Controllers/UsersController.cs (hunk 3 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs]: done

File: AccountingService/Controllers/UsersController.cs (hunk 4 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs:42]: done

File: AccountingService/Controllers/UsersController.cs (hunk 5 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs]: done

File: AccountingService/Controllers/UsersController.cs (hunk 9 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs]: done

File: AccountingService/Controllers/UsersController.cs (hunk 10 above)
Comment (berda97) [reply AccountingService/Controllers/UsersController.cs:90]: done

File: AccountingService/Controllers/UsersController.cs (hunk 1 above)
Comment (ghost) [reply AccountingService/Controllers/UsersController.cs]: Ther is one more.

File: AccountingService/Controllers/UsersController.cs [hunk 20]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
 
 
+                return Ok(user);
+            }
+            catch (Exception )
+            {
 
+                var error = new Error
+                {
+                    Message = "Failed to Get_users_buyID",
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: More meaningful comment: "Failed to get user"

File: AccountingService/Error.cs [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+    {
+        public string Message { get; set; }
+        public Code Code { get; set; }
+
+    }
+    public enum Code
+    {
+        Unknown,
+        Validation,
+
```
Comment (ghost) [inline AccountingService/Error.cs]: Remove new line.

File: AccountingService/Controllers/models/User.cs [hunk 22]
```diff
@@ -7,8 +7,9 @@ public class User
         public string LastName { get; set; }
         public string Address { get; set; }
         public double Salary { get; set; }
-        public string Position { get; set; }
         public string Currency { get; set; }
+        public string Position { get; set; }
+
```
Comment (ghost) [inline AccountingService/Controllers/models/User.cs:12]: Remove new line.

File: AccountingService/Controllers/UsersController.cs [hunk 23]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
             {
                 double netSalary = netSalaryService.Calculate(salary);
                 return Ok(new
                 {
                     Value = salary,
-                    Currency = currency,
-                    NetSalary = netSalary
+                    NetSalary = netSalary,
+                    Currency = currency
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove new line.

File: AccountingService/Controllers/UsersController.cs [hunk 24]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
+                }
+                 exchangeRate = exchangeRateService.GetCurrencyExchangeRate(currency);
+                
+            }
 
-            if (salary == 0)
+            catch (Exception )
             {
-                return NotFound();
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove new line.

File: AccountingService/Controllers/UsersController.cs [hunk 25]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
+                 salary = salaryConversionContext.User
+                    .Where(u => u.ID == id)
+                    .Select(u => u.Salary)
+                    .SingleOrDefault();
+                if(salary == 0)
+                {
+                    return NotFound();
+                }
+                 exchangeRate = exchangeRateService.GetCurrencyExchangeRate(currency);
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove new line.

File: AccountingService/Controllers/UsersController.cs [hunk 26]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
+                    var error = new Error
+                    {
+                        Message = "Failed to delete",
+                        Code = Code.Unknown
+                    };
+                    return BadGateway(error);
+                }
             }
-            return Ok();
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove new line.

File: AccountingService/Controllers/UsersController.cs (hunk 10 above)
Comment (ghost) [reply AccountingService/Controllers/UsersController.cs:90]: This is not done. Please re-add the check.

File: AccountingService/Controllers/UsersController.cs [hunk 27]
```diff
@@ -25,101 +22,164 @@ public UsersController(SalaryConversionContext context) : base()
...
+            try
             {
-                return BadGateway();
+                salaryConversionContext.User.Add(request);
+                int changes = salaryConversionContext.SaveChanges();
+                return Ok();
+            }
+            catch (Exception )
+            {
+
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: Remove new line.

File: AccountingService/Controllers/UsersController.cs [hunk 28]
```diff
@@ -10,135 +10,186 @@ namespace AccountingService.Controllers
...
-            salaryConversionContext.User.Add(request);
-            int changes = salaryConversionContext.SaveChanges();
-            if (changes == 0)
+            try
+            {
+                salaryConversionContext.User.Add(request);
+                int changes = salaryConversionContext.SaveChanges();
+                if (changes == 0)
+                {
+                    return NotFound();
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: We used to send here `BadGateway` not `NotFound`

File: AccountingService/Controllers/UsersController.cs [hunk 29]
```diff
@@ -10,135 +10,186 @@ namespace AccountingService.Controllers
...
-            int changes = salaryConversionContext.SaveChanges();
-            if (changes == 0)
+            try
             {
-                return BadGateway();
+                salaryConversionContext.User.Update(request);
+                int changes = salaryConversionContext.SaveChanges();
+                if (changes == 0)
+                {
+                    return NotFound();
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs]: `BadGateway` not `NotFound`

File: AccountingService/Controllers/UsersController.cs [hunk 30]
```diff
@@ -10,135 +10,186 @@ namespace AccountingService.Controllers
...
-                return BadGateway();
+                try
+                {
+                    var user = salaryConversionContext.User.Where(u => u.ID == id).SingleOrDefault();
+                   if (user == null)
+                    {
+                        return NotFound();
+                    }
+                    salaryConversionContext.User.Remove(user);
+                    int changes = salaryConversionContext.SaveChanges();
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:139]: Check is missing here

File: AccountingService/Controllers/UsersController.cs [hunk 31]
```diff
@@ -73,15 +76,20 @@ public IActionResult CreateUser([FromBody] User request)
                 int changes = salaryConversionContext.SaveChanges();
                 if (changes == 0)
                 {
-                    return NotFound();
+                    var fild_connection = new Error
```
Comment (ghost) [inline AccountingService/Controllers/UsersController.cs:78]: This is not how we name variables. You need to use `camelCase`.

**Review by ghost — APPROVED** [review]
LGTM
